pkg/master/controller/crdregistration/conversion.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crdregistration
+
+import (
+	"encoding/json"
+
+	"github.com/go-openapi/spec"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+// Convert_apiextensions_JSONSchemaProps_To_spec_Schema converts our internal OpenAPI v3 schema
+// (*apiextensions.JSONSchemaProps) to an OpenAPI v2 schema (*spec.Schema).
+func Convert_apiextensions_JSONSchemaProps_To_spec_Schema(in *apiextensions.JSONSchemaProps) (*spec.Schema, error) {
```

Comment:
Why does this use the internal types in the first place? This is a controller and controllers should use external types (the old controllers made this horrible mistake to be based on internal types, but we are working on fixing this).

Reply:
Updated. Thanks for pointing it out. I kept a conversion in `crdregistration_controller.go` to convert the internal `CustomResourceDefinitionSpec` to `v1beta1.CustomResourceDefinitionSpec`. The rest parts use the v1beta1 types.